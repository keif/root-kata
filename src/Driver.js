class Driver {
    constructor(driverName) {
        this._driverName = driverName;
        this._totalDrivingTimeInHrs = 0;
        this._totalDistInMiles = 0;
        this._milesPerHour = 0;
    }

    _calcTotalTimeInHrs(drivingTimes) {
        const { startHour, startMin, endHour, endMin } = drivingTimes;
        return endHour + endMin / 60 - (startHour + startMin / 60);
    }

    _convertTimeToNumber(drivingTimes) {
        return drivingTimes.split(":").map(num => Number(num));
    }

    getName() {
        return this._driverName;
    }

    getMilesPerHour() {
        return this._milesPerHour;
    }

    setMilesPerHour() {
        this._milesPerHour = Math.round(this._totalDistInMiles / this._totalDrivingTimeInHrs);
    }

    getTotalDistInMiles() {
        return this._totalDistInMiles;
    }

    setTotalDistInMiles(distInMiles) {
        this._totalDistInMiles += distInMiles;
        return this._totalDistInMiles;
    }

    getTotalTimeInHrs() {
        return this._totalDrivingTimeInHrs;
    }

    setTotalTimeInHrs(drivingTime) {
        const [startTime, endTime, ..._rest] = drivingTime;
        const [startHour, startMin] = this._convertTimeToNumber(startTime);
        const [endHour, endMin] = this._convertTimeToNumber(endTime);
        const drivingTimes = { startHour, startMin, endHour, endMin };
        this._totalDrivingTimeInHrs += this._calcTotalTimeInHrs(drivingTimes);
    }

    getRecord() {
        let driverDescription = `${this._driverName}: ${this._totalDistInMiles} miles`;

        const driverHasTraveled = this._totalDistInMiles;
        if (driverHasTraveled) {
            driverDescription += ` @ ${this._milesPerHour} mph`;
        }

        return driverDescription;
    }
}

export default Driver;
