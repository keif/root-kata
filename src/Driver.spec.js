import Driver from "./Driver";

let driver;

beforeAll(() => {
    let driverName = "Dan";
    driver = new Driver(driverName);
});

describe("Driver Class", () => {
    test('create a new driver with the name "Dan"', () => {
        expect(driver.getName()).toEqual("Dan");
    });

    test("should set Drivers driving time hours to 0.5", () => {
        driver.setTotalTimeInHrs(["07:15", "07:45"]);
        const driversTimeInHrs = driver.getTotalTimeInHrs();
        expect(driversTimeInHrs).toEqual(0.5);
    });

    test("should set Drivers distance in miles to 10", () => {
        driver.setTotalDistInMiles(10);
        const driversDistInMiles = driver.getTotalDistInMiles();
        expect(driversDistInMiles).toEqual(10);
    });

    test("should set drivers speed to 20", () => {
        driver.setMilesPerHour();
        const driversSpeed = driver.getMilesPerHour();
        expect(driversSpeed).toEqual(20);
    });

    test("should return a long descriptive string about the driver", () => {
        const driverDescription = driver.getRecord();
        const expectedDescription = "Dan: 10 miles @ 20 mph";
        expect(driverDescription).toEqual(expectedDescription);
    });
});
