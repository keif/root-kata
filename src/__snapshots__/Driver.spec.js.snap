// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Driver Class generates a driver object 1`] = `
Driver {
  "_driverName": "Dan",
  "_totalDrivingTime": 0,
  "_totalMilesDriven": 0,
  "_trips": Array [],
}
`;
